source/sessions/snapshot-sessions.rst

```diff
@@ -0,0 +1,265 @@
+
+Specification
+=============
+
+An application requests snapshot reads by creating a ``ClientSession``
+with options that specify that snapshot reads are desired. An
+application then passes the session as an argument to methods in the
+``MongoDatabase`` and ``MongoCollection`` classes. Read operations (find/aggregate/distinct) performed against
+that session will be read from the same snapshot.
+
+High level summary of the API changes for snapshot reads
+========================================================
+
+Snapshot reads are built on top of client sessions.
+
+Applications will start a new client session for snapshot reads like
+this:
+
+.. code:: typescript
+
+    options = new SessionOptions(snapshot = true);
+    session = client.startSession(options);
+
+All read operations performed using this session will be read from the same snapshot.
+
+If no value is provided for ``snapshot`` a value of false is
+implied.
+There are no MongoDatabase, MongoClient or MongoCollection API changes.
+
+SessionOptions changes
+======================
+
+``SessionOptions`` change summary
+
+.. code:: typescript
+
+    class SessionOptions {
+        Optional<bool> snapshot;
+
+        // other options defined by other specs
+    }
+
+In order to support snapshot reads a new property named
+``snapshot`` is added to ``SessionOptions``. Applications set
+``snapshot`` when starting a client session to indicate
+whether they want snapshot reads. All read operations performed
+using that client session will share the same snapshot.
+
+Each new member is documented below.
+
+snapshot
+--------
+
+Applications set ``snapshot`` when starting a session to
+indicate whether they want snapshot reads.
+
+Note that the ``snapshot`` property is optional. The default value of
+this property is false.
+
+Snapshot reads and causal consistency are mutually exclusive. Therefore if ``snapshot`` is set to true,
+``causalConsistency`` property is set to false. Client MUST throw an Error if both ``snapshot`` and ``causalConsistency`` are set to true.
+Snapshot reads are supported both on primaries and secondaries.
+
+ReadConcern changes
+===================
+
+``snapshot`` added to `ReadConcernLevel enumeration <../read-write-concern/read-write-concern.rst#read-concern>`_.`.
+
+Server Commands
+===============
+
+There are no new server commands related to snapshot reads. Instead,
+snapshot reads are implemented by:
+
+1. Saving the ``atClusterTime`` returned by 5.0+ servers for the first find/aggregate/distinct operation in a
+   private property ``snapshotTime`` of the ``ClientSession`` object. Drivers MUST save the ``atClusterTime``
+   in the ``ClientSession`` object.
+
+2. Passing that ``snapshotTime`` in the ``atClusterTime`` field of the ``readConcern`` field
+   for subsequent snapshot read operations (for find/aggregate/distinct commands).
+
+Server Command Responses
```

Comment:
When the driver starts a snapshot session but the server does not return the atClusterTime field the driver needs to raise a client side error indicating that server does not support snapshot reads. If we don't do this then snapshot reads can silently fail.

Reply:
False alarm. The problem was that my implementation neglected to set readConcern.level to ‘snapshot’. When the driver correctly sends ‘snapshot’ old servers will reject the request with this error:
```
{'operationTime': Timestamp(1624572085, 12), 'ok': 0.0, 'errmsg': 'read concern level snapshot is only valid in a transaction', 'code': 72, 'codeName': 'InvalidOptions', '$clusterTime': {'clusterTime': Timestamp(1624572085, 12), 'signature': {'hash': b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00', 'keyId': 0}}}
```